src/crypto/hsm/nxp/PersistentStorageOperationalKeystoreHSM.cpp

```diff
@@ -0,0 +1,288 @@
+    while (i < MAX_KEYID_SLOTS_FOR_FABRICS)
+    {
+        if (keyidFabIdMapping[i].keyId == 0 && keyidFabIdMapping[i].isPending == 0)
+        {
+            break;
+        }
+        i++;
+    }
+    return i;
+}
+
+void ResetPendingSlot(uint32_t slotId)
+{
+    if (slotId < MAX_KEYID_SLOTS_FOR_FABRICS)
+    {
+        keyidFabIdMapping[slotId].pkeyPair    = nullptr;
+        keyidFabIdMapping[slotId].keyId       = 0;
+        keyidFabIdMapping[slotId].fabricIndex = 0;
+        keyidFabIdMapping[slotId].isPending   = 0;
+    }
+}
+
+bool PersistentStorageOperationalKeystoreHSM::HasOpKeypairForFabric(FabricIndex fabricIndex) const
+{
+    uint8_t i = 0;
+    VerifyOrReturnError(IsValidFabricIndex(fabricIndex), false);
+
+    // If there was a pending keypair, then there's really a usable key
+    if (mIsPendingKeypairActive && (fabricIndex == mPendingFabricIndex) && (mPendingKeypair != nullptr))
+    {
+        ChipLogProgress(Crypto, "SE05x: HasOpKeypairForFabric ==> mPendingKeypair found");
+        return true;
+    }
+
+    while (i < MAX_KEYID_SLOTS_FOR_FABRICS)
+    {
+        if (keyidFabIdMapping[i].fabricIndex == fabricIndex)
+        {
+            ChipLogProgress(Crypto, "SE05x: HasOpKeypairForFabric ==> stored keyPair found");
+            return true;
+        }
+        i++;
+    }
+
+    ChipLogProgress(Crypto, "SE05x: HasOpKeypairForFabric ==> No key found");
+    return false;
+}
+
+CHIP_ERROR PersistentStorageOperationalKeystoreHSM::NewOpKeypairForFabric(FabricIndex fabricIndex,
+                                                                          MutableByteSpan & outCertificateSigningRequest)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    uint8_t slotId = getEmpytSlotId();
+
+    VerifyOrReturnError(slotId < MAX_KEYID_SLOTS_FOR_FABRICS, CHIP_ERROR_NO_MEMORY);
+    VerifyOrReturnError(IsValidFabricIndex(fabricIndex), CHIP_ERROR_INVALID_FABRIC_INDEX);
+
+    ChipLogProgress(Crypto, "SE05x: New Op Keypair for Fabric %02x", fabricIndex);
+
+    // Replace previous pending keypair, if any was previously allocated
+    ResetPendingKey();
+    if (mPendingKeypair != nullptr)
+    {
+        ResetPendingSlot(mPendingKeypair->GetKeyId() - FABRIC_SE05X_KEYID_START);
+    }
+
+    mPendingKeypair = Platform::New<Crypto::P256KeypairHSM>();
+    VerifyOrReturnError(mPendingKeypair != nullptr, CHIP_ERROR_NO_MEMORY);
+
+    // Key id is created as slotid + start offset of ops key id
+    mPendingKeypair->SetKeyId(FABRIC_SE05X_KEYID_START + slotId);
+
+    err = mPendingKeypair->Initialize();
+    VerifyOrReturnError(err == CHIP_NO_ERROR, CHIP_ERROR_NO_MEMORY);
+
+    mPendingFabricIndex = fabricIndex;
+
+    keyidFabIdMapping[slotId].isPending = 1;
+
+    size_t csrLength = outCertificateSigningRequest.size();
+    err              = mPendingKeypair->NewCertificateSigningRequest(outCertificateSigningRequest.data(), csrLength);
+    if (err != CHIP_NO_ERROR)
+    {
+        ResetPendingKey();
+        if (mPendingKeypair != nullptr)
```

Comment:
Given this code, mPendingKeypair is always null here.  Shouldn't the `ResetPendingSlot` just happen inside `ResetPendingKey`?

Reply:
Yes. Fixed this. Moved ResetPendingSlot to ResetPendingKey